src/main/java/com/redhat/devtools/lsp4ij/features/selectionRange/LSPCodeBlockStatementGroupSelectioner.java

```diff
@@ -0,0 +1,125 @@
+        Document document = editor.getDocument();
+        if (isEmpty(document, editorText, currentLineNumber)) {
+            return null;
+        }
+
+        // The current statement group contains all non-empty lines within the code block from the start offset
+        int codeBlockStartLineNumber = document.getLineNumber(codeBlockRange.getStartOffset());
+        int codeBlockEndLineNumber = document.getLineNumber(codeBlockRange.getEndOffset());
+        if ((codeBlockStartLineNumber >= currentLineNumber) || (codeBlockEndLineNumber <= currentLineNumber)) {
+            return null;
+        }
+
+        int statementGroupStartLineNumber = currentLineNumber;
+        while ((statementGroupStartLineNumber > codeBlockStartLineNumber) &&
+               !isEmpty(document, editorText, statementGroupStartLineNumber - 1)) {
+            int pendingStatementGroupStartLineNumber = statementGroupStartLineNumber - 1;
+
+            // If this line looks like it might end a nested code block, confirm that to be the case or not, and if so,
+            // add all lines in that nested code block
+            int currentLineStartOffset = document.getLineStartOffset(statementGroupStartLineNumber);
+            int currentLineEndOffset = document.getLineEndOffset(statementGroupStartLineNumber);
+            while ((currentLineStartOffset <= currentLineEndOffset) && Character.isWhitespace(editorText.charAt(currentLineStartOffset))) {
+                currentLineStartOffset++;
+            }
+            if (LSPCodeBlockUtils.isCodeBlockEndChar(file, editorText.charAt(currentLineStartOffset))) {
+                TextRange nestedCodeBlockRange = LSPCodeBlockProvider.getCodeBlockRange(editor, file, currentLineStartOffset - 1);
+                if ((nestedCodeBlockRange != null) && codeBlockRange.contains(nestedCodeBlockRange) && !Objects.equals(codeBlockRange, nestedCodeBlockRange)) {
+                    int nestedCodeBlockStartLineNumber = document.getLineNumber(nestedCodeBlockRange.getStartOffset());
+                    if ((nestedCodeBlockStartLineNumber > codeBlockStartLineNumber) && (nestedCodeBlockStartLineNumber < statementGroupStartLineNumber)) {
+                        pendingStatementGroupStartLineNumber = nestedCodeBlockStartLineNumber;
+                    }
+                }
+            }
+
+            statementGroupStartLineNumber = pendingStatementGroupStartLineNumber;
+        }
+        statementGroupStartLineNumber = Math.max(statementGroupStartLineNumber, codeBlockStartLineNumber + 1);
+
+        int statementGroupEndLineNumber = currentLineNumber;
+        while ((statementGroupEndLineNumber < codeBlockEndLineNumber) &&
+               !isEmpty(document, editorText, statementGroupEndLineNumber + 1)) {
+            int pendingStatementGroupEndLineNumber = statementGroupEndLineNumber + 1;
+
+            // If this line looks like it might start a nested code block, confirm that to be the case or not, and if
```

Comment:
We do the exact same thing when searching forwards for additional statements in the group, but instead looking for lines that end with a code block start character.

This does mean that there are, by necessity, multiple calls to `LSPCodeBlockProvider.getCodeBlockRange()` to get the full statement group when one or more statements in that group have nested code blocks, but it ends up being `n+1` calls overall where `n` is the number of statements in the group with nested code blocks.